AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/AWSS3StoragePlugin+StorageBucket.swift

```diff
@@ -0,0 +1,74 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+@_spi(InternalAmplifyConfiguration) import Amplify
+import Foundation
+
+extension AWSS3StoragePlugin {
+    /// Returns a AWSS3StorageServiceBehavior instance for the given StorageBucket
+    func storageService(for bucket: (any StorageBucket)?) throws -> AWSS3StorageServiceBehavior {
+        guard let bucket else {
+            // When no bucket is provided, use the default one
+            return defaultStorageService
+        }
+
+        let bucketInfo = try bucketInfo(from: bucket)
+        guard let storageService = storageServicesByBucket[bucketInfo.bucketName] else {
+            // If no service was found for the bucket, create one
+            let storageService = try createStorageService(
+                authService: authService,
+                bucketInfo: bucketInfo
+            )
+            storageServicesByBucket[bucketInfo.bucketName] = storageService
+            return storageService
+        }
+
+        return storageService
+    }
+
+    /// Returns a AWSS3StorageServiceProvider callback for the given StorageBucket
+    func storageServiceProvider(for bucket: (any StorageBucket)?) -> AWSS3StorageServiceProvider {
+        let storageServiceResolver: () throws -> AWSS3StorageServiceBehavior = { [weak self] in
+            guard let self = self else {
+                throw StorageError.unknown("AWSS3StoragePlugin was deallocated", nil)
+            }
+            return try self.storageService(for: bucket)
+        }
+        return storageServiceResolver
+    }
+
+    /// Returns a valid `BucketInfo` instance from the given StorageBucket
+    private func bucketInfo(from bucket: any StorageBucket) throws -> BucketInfo {
+        switch bucket {
+        case let outputsBucket as OutputsStorageBucket:
+            guard let additionalBucketsByName else {
```

Comment:
based on the approved API design, on Android, an `InvalidStorageBucketException` is returned for error cases; was there any consideration to provided a specific exception for Swift?

Reply:
The approved API design has `StorageError.validation` being thrown for Swift, not a new error.
This is consistent to what happens when an invalid `StoragePath` is provided, so I didn't explore introducing a new error.